File: src/gui/Game.java
package gui;

import static utils.InputHandler.*;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

import entity.bullet.Bullet;
import entity.enemy.Enemy;
import entity.tower.Tower;
import javafx.event.ActionEvent;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.control.Button;
import javafx.scene.image.Image;
import javafx.scene.input.MouseEvent;
import javafx.scene.paint.Color;
import level.Level;

/**
 * Created by zomby on 5/10/17 @ 4:20 PM.
 */
public class Game extends Canvas {

	public static int TILE_SIZE = 64;
	private final int GRID_WIDTH = 16;
	private final int GRID_HEIGHT = 10;

	private Image[] gridSet;

	private final int[][] grid = new int[][] { { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
			{ 0, 0, 0, 5, 1, 1, 1, 1, 1, 1, 1, 6, 0, 0, 0, 0 }, { 0, 0, 0, 4, 9, 3, 3, 3, 3, 3, 10, 2, 0, 0, 0, 0 },
			{ 0, 0, 0, 4, 2, 0, 0, 0, 0, 0, 4, 2, 0, 0, 0, 0 }, { 0, 0, 0, 4, 2, 0, 0, 5, 1, 1, 11, 2, 0, 0, 0, 0 },
			{ 0, 0, 0, 4, 2, 0, 0, 4, 9, 3, 3, 7, 0, 0, 0, 0 }, { 0, 0, 0, 4, 2, 0, 0, 4, 2, 0, 0, 0, 0, 0, 0, 0 },
			{ 1, 1, 1, 11, 2, 0, 0, 4, 2, 0, 0, 0, 0, 0, 0, 0 }, { 3, 3, 3, 3, 7, 0, 0, 4, 12, 1, 1, 1, 1, 1, 1, 1 },
			{ 0, 0, 0, 0, 0, 0, 0, 8, 3, 3, 3, 3, 3, 3, 3, 3 } };

	private final char[][] route = new char[][] {
			{ 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X' },
			{ 'X', 'X', 'X', 'E', 'X', 'X', 'X', 'X', 'X', 'X', 'S', 'X', 'X', 'X', 'X' },
			{ 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X' },
			{ 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X' },
			{ 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'S', 'X', 'X', 'W', 'X', 'X', 'X', 'X' },
			{ 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X' },
			{ 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X' },
			{ 'E', 'X', 'X', 'N', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X' },
			{ 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'E', 'X', 'X', 'X', 'X', 'X', 'X', 'X' } };

	private List<Tower> towers;
	private List<Enemy> enemies;
	private List<Bullet> bullets;

	private int mx, my, money = 50, levelNum = 1;
	private Tower selectedTower;
	private Button levelButton;

	private boolean contained = true;

	public Game(int width, int height) {
		// TODO
		super(width, height);

		towers = Collections.synchronizedList(new ArrayList<>());
		enemies = Collections.synchronizedList(new ArrayList<>());
		bullets = Collections.synchronizedList(new ArrayList<>());

		init();
	}

	public void setLevelButton(Button levelButton) {
		this.levelButton = levelButton;
		this.levelButton.setOnAction((ActionEvent e) -> {
			new Thread(new Level(levelNum++, this)).start();
			this.levelButton.setVisible(false);
		});
	}

	public void onLevelDone() {
		levelButton.setVisible(true);
	}

	public int getMoney() {
		return money;
	}

	public void buyTower(int cost) {
		money -= cost;
	}

	public void setSelectedTower(Tower t) {
		selectedTower = t;
	}

	public void addEnemy(Enemy simpleEnemy) {
		enemies.add(simpleEnemy);
	}

	private void init() {
		// TODO

		gridSet = new Image[13];
		gridSet[0] = new Image("file:img/PNG/Retina/towerDefense_tile162.png");
		gridSet[1] = new Image("file:img/PNG/Retina/towerDefense_tile254.png");
		gridSet[2] = new Image("file:img/PNG/Retina/towerDefense_tile230.png");
		gridSet[3] = new Image("file:img/PNG/Retina/towerDefense_tile208.png");
		gridSet[4] = new Image("file:img/PNG/Retina/towerDefense_tile232.png");
		gridSet[5] = new Image("file:img/PNG/Retina/towerDefense_tile210.png");
		gridSet[6] = new Image("file:img/PNG/Retina/towerDefense_tile211.png");
		gridSet[7] = new Image("file:img/PNG/Retina/towerDefense_tile234.png");
		gridSet[8] = new Image("file:img/PNG/Retina/towerDefense_tile233.png");
		gridSet[9] = new Image("file:img/PNG/Retina/towerDefense_tile207.png");
		gridSet[10] = new Image("file:img/PNG/Retina/towerDefense_tile209.png");
		gridSet[11] = new Image("file:img/PNG/Retina/towerDefense_tile255.png");
		gridSet[12] = new Image("file:img/PNG/Retina/towerDefense_tile253.png");

		setOnMouseMoved((MouseEvent e) -> {
			mx = (int) e.getX();
			my = (int) e.getY();
		});

		setOnMouseExited((MouseEvent e) -> {
			contained = false;
		});

		setOnMouseEntered((MouseEvent e) -> {
			contained = true;
		});

		setOnMouseClicked((MouseEvent e) -> {
			selectedTower = handleMouseClick(e, grid, towers, selectedTower, this);
		});

	}

	public void repaint() {
		// TODO
		GraphicsContext gc = getGraphicsContext2D();

		gc.clearRect(0, 0, TILE_SIZE * GRID_WIDTH, TILE_SIZE * GRID_HEIGHT);

		// DRAW BACKGROUND MAP
		// -------------------
		for (int i = 0; i < GRID_WIDTH; i++) {
			for (int j = 0; j < GRID_HEIGHT; j++) {
				gc.drawImage(gridSet[grid[j][i]], i * TILE_SIZE, j * TILE_SIZE, TILE_SIZE, TILE_SIZE);
			}
		}
		// -------------------

		// DRAW TOWERS
		synchronized (towers) {
			for (Tower t : towers) {
				t.draw(gc);
			}
		}

		// DRAW ENEMIES
		synchronized (enemies) {
			for (Enemy e : enemies) {
				e.draw(gc);
			}
		}

		// DRAW BULETS
		synchronized (bullets) {
			for (Bullet b : bullets) {
				b.draw(gc);
			}
		}

		// DRAW THE SELECTED TOWER
		if (selectedTower != null && contained) {
			if (grid[my / TILE_SIZE][mx / TILE_SIZE] == 0 && !collides(towers, mx, my)) {
				selectedTower.draw(gc, mx, my);

				int range = selectedTower.getRange();
				gc.setStroke(Color.RED);
				gc.strokeOval(mx - range, my - range, range * 2, range * 2);
			} else {
				gc.fillOval(mx, my, 5, 5);
			}
		}

		gc.fillText("Money: " + money, 100, 100);

	}

	public void update() {
		synchronized (bullets) {
			Iterator<Bullet> bIterator = bullets.iterator();

			while (bIterator.hasNext()) {
				Bullet b = bIterator.next();

				b.move();

				if (b.collided()) {
					b.doDamage();
					bIterator.remove();
					b = null;
				}
			}
		}

		synchronized (enemies) {
			Iterator<Enemy> eIterator = enemies.iterator();

			while (eIterator.hasNext()) {
				Enemy e = eIterator.next();

				if (e.isDead()) {
					eIterator.remove();
					money += e.getReward();
					e = null;
					continue;
				}

				e.move(route);
			}
		}

		for (Tower t : towers) {

			Enemy frontEnemy = null;

			for (Enemy e : enemies) {
				if (e.distance(t) < t.getRange()) {
					if (frontEnemy == null) {
						frontEnemy = e;
					} else if (e.getDistanceTraveled() > frontEnemy.getDistanceTraveled()) {
						frontEnemy = e;
					}
				}
			}
			
			if (frontEnemy != null && t.canFire()) {
				bullets.add(t.fire(frontEnemy));
			}
			
			t.updateAngle();
		}
	}
}

File: src/gui/Game.kt
package gui

import entity.bullet.Bullet
import entity.enemy.Enemy
import entity.tower.Tower
import javafx.event.ActionEvent
import javafx.scene.canvas.Canvas
import javafx.scene.control.Button
import javafx.scene.image.Image
import javafx.scene.input.MouseEvent
import javafx.scene.paint.Color
import javafx.scene.text.Font
import level.Level
import utils.InputHandler.collides
import utils.InputHandler.handleMouseClick
import java.util.*

class Game(width: Int, height: Int) : Canvas(width.toDouble(), height.toDouble()) {
    private val GRID_WIDTH = 16
    private val GRID_HEIGHT = 10

    private var gridSet: Array<Image?> = arrayOfNulls(13)

    private val grid = arrayOf(intArrayOf(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), intArrayOf(0, 0, 0, 5, 1, 1, 1, 1, 1, 1, 1, 6, 0, 0, 0, 0), intArrayOf(0, 0, 0, 4, 9, 3, 3, 3, 3, 3, 10, 2, 0, 0, 0, 0), intArrayOf(0, 0, 0, 4, 2, 0, 0, 0, 0, 0, 4, 2, 0, 0, 0, 0), intArrayOf(0, 0, 0, 4, 2, 0, 0, 5, 1, 1, 11, 2, 0, 0, 0, 0), intArrayOf(0, 0, 0, 4, 2, 0, 0, 4, 9, 3, 3, 7, 0, 0, 0, 0), intArrayOf(0, 0, 0, 4, 2, 0, 0, 4, 2, 0, 0, 0, 0, 0, 0, 0), intArrayOf(1, 1, 1, 11, 2, 0, 0, 4, 2, 0, 0, 0, 0, 0, 0, 0), intArrayOf(3, 3, 3, 3, 7, 0, 0, 4, 12, 1, 1, 1, 1, 1, 1, 1), intArrayOf(0, 0, 0, 0, 0, 0, 0, 8, 3, 3, 3, 3, 3, 3, 3, 3))

    private val route = arrayOf(charArrayOf('X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X'), charArrayOf('X', 'X', 'X', 'E', 'X', 'X', 'X', 'X', 'X', 'X', 'S', 'X', 'X', 'X', 'X'), charArrayOf('X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X'), charArrayOf('X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X'), charArrayOf('X', 'X', 'X', 'X', 'X', 'X', 'X', 'S', 'X', 'X', 'W', 'X', 'X', 'X', 'X'), charArrayOf('X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X'), charArrayOf('X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X'), charArrayOf('E', 'X', 'X', 'N', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X', 'X'), charArrayOf('X', 'X', 'X', 'X', 'X', 'X', 'X', 'E', 'X', 'X', 'X', 'X', 'X', 'X', 'X'))

    private val towers: MutableList<Tower> = Collections.synchronizedList(ArrayList<Tower>())
    private val enemies: MutableList<Enemy> = Collections.synchronizedList(ArrayList<Enemy>())
    private val bullets: MutableList<Bullet> = Collections.synchronizedList(ArrayList<Bullet>())

    private var mx = 0
    private var my = 0

    internal var money = 50
        get set

    private var selectedTower: Tower? = null
    private var levelButton: Button? = null

    private var contained = true

    init {
        gridSet[0] = Image("file:img/PNG/Retina/towerDefense_tile162.png")
        gridSet[1] = Image("file:img/PNG/Retina/towerDefense_tile254.png")
        gridSet[2] = Image("file:img/PNG/Retina/towerDefense_tile230.png")
        gridSet[3] = Image("file:img/PNG/Retina/towerDefense_tile208.png")
        gridSet[4] = Image("file:img/PNG/Retina/towerDefense_tile232.png")
        gridSet[5] = Image("file:img/PNG/Retina/towerDefense_tile210.png")
        gridSet[6] = Image("file:img/PNG/Retina/towerDefense_tile211.png")
        gridSet[7] = Image("file:img/PNG/Retina/towerDefense_tile234.png")
        gridSet[8] = Image("file:img/PNG/Retina/towerDefense_tile233.png")
        gridSet[9] = Image("file:img/PNG/Retina/towerDefense_tile207.png")
        gridSet[10] = Image("file:img/PNG/Retina/towerDefense_tile209.png")
        gridSet[11] = Image("file:img/PNG/Retina/towerDefense_tile255.png")
        gridSet[12] = Image("file:img/PNG/Retina/towerDefense_tile253.png")

        setOnMouseMoved { e: MouseEvent ->
            mx = e.x.toInt()
            my = e.y.toInt()
        }

        setOnMouseExited { _: MouseEvent -> contained = false }

        setOnMouseEntered { _: MouseEvent -> contained = true }

        setOnMouseClicked { e: MouseEvent -> selectedTower = handleMouseClick(e, grid, towers, selectedTower, this) }
    }

    fun setLevelButton(levelButton: Button) {
        this.levelButton = levelButton
        this.levelButton?.setOnAction { _: ActionEvent ->
            bullets.clear()
            Thread(Level(this)).start()
            this.levelButton?.isVisible = false
        }
    }

    fun onLevelDone() {
        levelButton!!.isVisible = true
    }

    fun buyTower(cost: Int) {
        money -= cost
    }

    fun setSelectedTower(t: Tower) {
        selectedTower = t
    }

    fun addEnemy(simpleEnemy: Enemy) {
        enemies.add(simpleEnemy)
    }

    fun repaint() {
        val gc = graphicsContext2D

        gc.clearRect(0.0, 0.0, (TILE_SIZE * GRID_WIDTH).toDouble(), (TILE_SIZE * GRID_HEIGHT).toDouble())

        // DRAW BACKGROUND MAP
        // -------------------
        for (i in 0..GRID_WIDTH - 1) {
            for (j in 0..GRID_HEIGHT - 1) {
                gc.drawImage(gridSet[grid[j][i]], (i * TILE_SIZE).toDouble(), (j * TILE_SIZE).toDouble(), TILE_SIZE.toDouble(), TILE_SIZE.toDouble())
            }
        }
        // -------------------

        // DRAW TOWERS
        synchronized(towers) {
            for (t in towers) {
                t.draw(gc)
            }
        }

        // DRAW ENEMIES
        synchronized(enemies) {
            for (e in enemies) {
                e.draw(gc)
            }
        }

        // DRAW BULETS
        synchronized(bullets) {
            for (b in bullets) {
                b.draw(gc)
            }
        }

        // DRAW THE SELECTED TOWER
        if (selectedTower != null && contained) {
            if (grid[my / TILE_SIZE][mx / TILE_SIZE] == 0 && !collides(towers, mx, my)) {
                selectedTower!!.draw(gc, mx, my)

                val range = selectedTower!!.range
                gc.stroke = Color.RED
                gc.strokeOval((mx - range).toDouble(), (my - range).toDouble(), (range * 2).toDouble(), (range * 2).toDouble())
            } else {
                gc.fillOval(mx.toDouble(), my.toDouble(), 5.0, 5.0)
            }
        }

        gc.font = FONT
        gc.fillText("Money: " + money, 10.0, 100.0)
    }

    fun update() {
        synchronized(bullets) {
            val bIterator = bullets.iterator()

            while (bIterator.hasNext()) {
                val b: Bullet = bIterator.next()

                b.move()

                if (b.collided()) {
                    b.doDamage()
                    bIterator.remove()
                } else if (b.offScreen()) {
                    bIterator.remove()
                }
            }
        }

        synchronized(enemies) {
            val eIterator = enemies.iterator()

            while (eIterator.hasNext()) {
                val e: Enemy = eIterator.next()

                if (e.isDead) {
                    eIterator.remove()
                    money += e.reward
                    continue
                }

                e.move(route)
            }
        }

        for (t in towers) {

            var frontEnemy: Enemy? = null

            for (e in enemies) {
                if (e.distance(t) < t.range) {
                    if (frontEnemy == null) {
                        frontEnemy = e
                    } else if (e.distanceTraveled > frontEnemy.distanceTraveled) {
                        frontEnemy = e
                    }
                }
            }

            if (frontEnemy != null && t.canFire) {
                bullets.add(t.fire(frontEnemy))
            }

            t.updateAngle()
        }
    }

    companion object {
        val FONT = Font("KenVector Future Regular", 22.0)
        val TILE_SIZE = 64
    }
}
